ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -175,6 +198,167 @@ def _map_sample(self, sample, exclude_fields):
         return sample
 
 
+@dataclasses.dataclass
+class TimestampGenerationTask:
+    """
+    This task is used to generate timestamps for the text transcription.
+    It uses the Montreal Forced Aligner (MFA) to align the text with the audio. The result is a
+    list of timestamps for each word in the text transcription. The timestamps are stored in a new
+    column, in a list of dict format:
+        [ {"start": float in seconds, "end": float in seconds, "text": first word str}, ... ]
+    """
+
+    # Jinja template for the text transcription that needs to be aligned
+    template: str = simple_parsing.field(alias="-T")
+    # The accoustic model to use for MFA alignment.
+    # Make sure the dictionary and acoustic model are installed. See just install_mfa for an example (English).
+    # Model index: https://mfa-models.readthedocs.io/en/latest/acoustic/index.html
+    # For many languages there exists a {language}_mfa model that you can use, e.g. "english_mfa"
+    mfa_acoustic_model: str = simple_parsing.field(alias="-m")
+    # The dictionary to use for MFA alignment. Defaults to the same name as the acoustic model.
+    mfa_dictionary: Optional[str] = simple_parsing.field(default=None, alias="-d")
+    audio_column_name: str = simple_parsing.field(default="audio", alias="-a")
+    sample_rate: int = simple_parsing.field(default=16000, alias="-r")
+    # The column name to store the timestamps in
+    timestamp_column_name: str = simple_parsing.field(default="timestamps", alias="-ts")
+    aligned_ratio_check: float = simple_parsing.field(default=0.95, alias="-ar")
+
+    def __post_init__(self):
+        if self.mfa_dictionary is None:
+            self.mfa_dictionary = self.mfa_acoustic_model
+
+        try:
+            # Make sure the MFA environment is installed
+            subprocess.run(["conda", "run", "-n", MFA_ENV_NAME, "echo"], check=True)
+        except subprocess.CalledProcessError:
+            raise Exception(
+                "Please install the MFA environment using `just install_mfa` first."
+            )
+
+        if self.template.startswith("@"):
+            with open(self.template[1:], "r") as template_file:
+                self.template = template_file.read()
+
+    def map_split(
+        self,
+        ds_split: datasets.Dataset,
+        num_proc: int,
+        writer_batch_size: int,
+        exclude_fields: List[str],
+    ) -> datasets.Dataset:
+        # 0. create a temp directory to store the audio and text files
+        _temp_dir = tempfile.TemporaryDirectory()
+        self.temp_dir = _temp_dir.name
+        os.makedirs(self.temp_dir, exist_ok=True)
+
+        # 1. copy all audio-text pairs into the temp directory
+        ds_split.map(
```

Comment:
Can you elaborate by what's redundant?
If you mean why we store the audio and text to file, that's just how MFA expects the inputs. I didn't find a way to invoke it from Python and let it read from the dataset, hence the awkward: store, compute, retrieve, cleanup loop.

The larger datasets use your chunking and retry machinery by default, so that's been working fine for Common Voice and has allowed for retrying failed invocations.

Reply:
Ah okay if that's how MFA works, then ig there's no other way around it. 

I was concerned that you might run out of storage memory, so was thinking it might be good to save the audio in chunks and delete the chunk once its been processed?

Reply:
Oh okay yes, the temp directory's cleanup command will run an `rmtree` and removes everything inside.

Reply:
But that reminds me: I can put this in a `with` statement to make sure the directory is removed even if an error gets raised.